src/dotnet/ToolPackage/ToolPackageObtainer.cs

```diff
@@ -42,6 +41,79 @@ IProjectRestorer projectRestorer
             string targetframework = null,
             string source = null,
             string verbosity = null)
+        {
+            var stageDirectory = _toolsPath.WithSubDirectories(".stage", Path.GetRandomFileName());
+
+            var resourceManager = new TransactionResourceManager(
+                commit: () =>
+                {
+                    Directory.Move(
```

Comment:
Same general comments as above.  If this move or delete operation fails, the transaction will not be rolled back and we'll be left in an inconsistent state.

Reply:
Yes and it should not fail. If there is uncontrollable case like control-c, nothing can prevent this inconsistency.  We can only rely on `install --force` to correct it.

Reply:
I think it is common pattern to use Move as a confirm for file system, there is no real commit as in database.

Also, it is way more complicated if we don't trust move. Rollback need to be different before move and after move. Also, how do we trust Delete then?

Reply:
But unlike the "process killed" inconsistency (one we can never protect ourselves from), this is a preventable one; if we failed to prepare the transaction (i.e. because we couldn't move from staging), then the transaction should be rolled back.  The way this is implemented now won't roll back the transaction.

I like to also speak towards the "correctness" of the implementation.  Not that we'll ever do so, but if we create a transaction and perform the same operation twice (such as install the same package or create the same shim), then I'd expect the second operation to fail and rollback the entire transaction.  Right now this won't happen because both `Prepare` calls will succeed, but both `Commit` calls will not.

Reply:
> this is a preventable one

In prepare, the existence of a folder with the same name already checked. I don't know what else can cause this to fail. Other than issue like process killed, permission denied (it is a folder dotnet created)

Again, how do you prevent this happen when Move does not considered commit.

There is an existing package
Download package to staging
Diectory.Move failed, due to some reason
Rollback, delete .dotnet/packageid, so the existing proper package get deleted.

> Right now this won't happen because both Prepare calls will succeed, but both Commit calls will not.

Prepare **will** fail, it checks the folder existence or not

Reply:
Right, but what about:

```c#
using (var scope = new TransactionScope())
{
    var obtainer = new ToolPackageObtainer(...);
    obtainer.ObtainAndReturnExecutablePath("foo");
    obtainer.ObtainAndReturnExecutablePath("foo");
    scope.Complete();
}
```

The order of operations is:

1. Prepare: check for existence of "foo" package directory...succeeds.
2. Prepare: check for existence of "foo" package directory...succeeds.
3. Commit: move directory...succeeds.
4. Commit: move directory...fails and transaction is not rolled back.

Now I'm not saying we need to support this, I'm just arguing from a correctness standpoint, I'd expect this to fail and rollback as if I had inserted two rows into a database with the same primary key in the same transaction.  Neither insert should remain.

If the `Prepare` call did the move, then the subsequent Prepare call would detect it, fail, and rollback the entire transaction.

Reply:
Decide on empty commit and empty prepare. Use a non generic IEnlistmentNotification to track all the state of transaction, including it is before move or after move

Reply:
Changed to what we decide on, please review